src/coreclr/src/vm/assemblynative.cpp

```diff
@@ -31,55 +31,48 @@
 #include "../binder/inc/bindertracing.h"
 #include "../binder/inc/clrprivbindercoreclr.h"
 
-FCIMPL5(Object*, AssemblyNative::Load, AssemblyNameBaseObject* assemblyNameUNSAFE,
-        AssemblyBaseObject* requestingAssemblyUNSAFE,
-        StackCrawlMark* stackMark,
-        CLR_BOOL fThrowOnFileNotFound,
-        AssemblyLoadContextBaseObject *assemblyLoadContextUNSAFE)
+/* static */
+void QCALLTYPE AssemblyNative::Load(QCall::ObjectHandleOnStack assemblyName, QCall::ObjectHandleOnStack requestingAssembly, QCall::StackCrawlMarkHandle stackMark,
+                                    BOOL fThrowOnFileNotFound, QCall::ObjectHandleOnStack assemblyLoadContext, QCall::ObjectHandleOnStack retAssembly)
 {
-    FCALL_CONTRACT;
-
-    struct _gc
-    {
-        ASSEMBLYNAMEREF        assemblyName;
-        ASSEMBLYREF            requestingAssembly;
-        ASSEMBLYREF            rv;
-        ASSEMBLYLOADCONTEXTREF assemblyLoadContext;
-    } gc;
-
-    gc.assemblyName        = (ASSEMBLYNAMEREF)        assemblyNameUNSAFE;
-    gc.requestingAssembly  = (ASSEMBLYREF)            requestingAssemblyUNSAFE;
-    gc.rv                  = NULL;
-    gc.assemblyLoadContext = (ASSEMBLYLOADCONTEXTREF) assemblyLoadContextUNSAFE;
+    QCALL_CONTRACT;
 
-    HELPER_METHOD_FRAME_BEGIN_RET_PROTECT(gc);
+    BEGIN_QCALL;
 
-    if (gc.assemblyName == NULL)
+    if (*assemblyName.m_ppObject == NULL)
+    {
         COMPlusThrow(kArgumentNullException, W("ArgumentNull_AssemblyName"));
-
+    }
     ACQUIRE_STACKING_ALLOCATOR(pStackingAllocator);
 
     DomainAssembly * pParentAssembly = NULL;
     Assembly * pRefAssembly = NULL;
+    ICLRPrivBinder *pBinderContext = NULL;
 
-    INT_PTR ptrLoadContextBinder = (gc.assemblyLoadContext != NULL) ? gc.assemblyLoadContext->GetNativeAssemblyLoadContext() : NULL;
-
-    if(gc.assemblyName->GetSimpleName() == NULL)
+    if (*assemblyLoadContext.m_ppObject != NULL)
+    {
+        GCX_COOP();
+        ASSEMBLYLOADCONTEXTREF * pAssemblyLoadContextRef = reinterpret_cast<ASSEMBLYLOADCONTEXTREF *>(assemblyLoadContext.m_ppObject);
+        INT_PTR nativeAssemblyLoadContext = (*pAssemblyLoadContextRef)->GetNativeAssemblyLoadContext();
+        pBinderContext = reinterpret_cast<ICLRPrivBinder*>(nativeAssemblyLoadContext);
+    }
+    ASSEMBLYNAMEREF * pAssemblyNameRef = reinterpret_cast<ASSEMBLYNAMEREF *>(assemblyName.m_ppObject);
+    if ((*pAssemblyNameRef)->GetSimpleName() == NULL)
```

Comment:
This needs to be under `GCX_COOP`. ASSEMBLYNAMEREF is object reference.

Reply:
It may be easiest if you have just one `GCX_COOP` for the whole function. There are a lot of places that work with object references.